haskellPackages: stackage LTS 24.12 -> LTS 24.16

wolfgangwalther opened the pull request:
### This Merge

This PR is the regular merge of the `haskell-updates` branch into `staging`.

This branch is being continually built and tested by hydra at https://hydra.nixos.org/jobset/nixpkgs/haskell-updates. You may be able to find an up-to-date Hydra build report at [cdepillabout/nix-haskell-updates-status](https://github.com/cdepillabout/nix-haskell-updates-status).

We roughly aim to merge these `haskell-updates` PRs at least once every two weeks. See the @NixOS/haskell [team calendar](https://cloud.maralorn.de/apps/calendar/p/H6migHmKX7xHoTFa) for who is currently in charge of this branch.

### haskellPackages Workflow Summary

Our workflow is currently described in [`pkgs/development/haskell-modules/HACKING.md`](https://github.com/NixOS/nixpkgs/blob/haskell-updates/pkgs/development/haskell-modules/HACKING.md).

The short version is this:
* We regularly update the Stackage and Hackage pins on `haskell-updates` (normally at the beginning of a merge window).
* The community fixes builds of Haskell packages on that branch.
* We aim at at least one merge of `haskell-updates` into `staging` every two weeks.
* We only do the merge if the [`mergeable`](https://hydra.nixos.org/job/nixpkgs/haskell-updates/mergeable) job is succeeding on hydra.
* If a [`maintained`](https://hydra.nixos.org/job/nixpkgs/haskell-updates/maintained) package is still broken at the time of merge, we will only merge if the maintainer has been pinged 7 days in advance. (If you care about a Haskell package, become a maintainer!)

More information about Haskell packages in nixpkgs can be found [in the nixpkgs manual](https://nixos.org/manual/nixpkgs/unstable/#haskell).

---

This is the follow-up to #445213. Come to [#haskell:nixos.org](https://matrix.to/#/#haskell:nixos.org) if you have any questions.

wolfgangwalther committed:
haskell.compiler.ghc9101: drop

Latest 9.10.x minor release is 9.10.2, which is also in Stackage 24.8.

Thus, dropping according to the GHC Deprecation Policy.

(cherry picked from commit ac5228c6c02590a0d05afad272e1598385ee9680)

wolfgangwalther committed:
haskell.compiler.ghc9121: drop

Latest 9.12.x minor release is 9.12.2, which is also in Stackage Nightly
2025-09-01.

Thus, dropping according to the GHC Deprecation Policy.

emilazy committed:
haskell.compiler.ghc{948,967,984,9102,9103,9121,9122}: backport patches for LLVM support

LLVM 12–17 have been dropped for Nixpkgs 25.11. As discussed recently
on Matrix, this backports upstream changes to allow the use of
LLVM 20 for all GHC versions from 9.4.8 onward.

I looked over GHC commits mentioning LLVM since the release of 9.4.8,
and read the discussions and issues around the relevant bumps, and
attempted to be quite thorough, but I obviously cannot guarantee that
this is wholly comprehensive. It seems like upstream generally bumps
the upper bound on the basis of “it builds successfully for me”,
with specific adaptations for new versions being fairly uncommon and
only coming for obvious build blockers or reactively in response to
bug reports. I have backported both kinds of changes here.

For some commits, trivial conflict resolutions and adaptations
were required. It would be possible to pass the affected files to
`fetchpatch` as `excludes` and keep smaller fix‐up patches in tree in
some cases, but I opted to keep it simple and vendor complete backport
patches instead. I did not attempt to backport every single change to
the LLVM backend, only those that seemed directly relevant to support
for newer versions; if you’d get the same issue with the older LLVM,
that’s just a GHC bug.

These changes should actually make it easier to cross‐compile for
new architectures, as more recent LLVMs will have better support for
newer platforms, and it will be easier to backport GHC changes to
enable new platforms with less drift in the backend.

These patches do result in two breaking changes. Firstly, the minimum
LLVM version is bumped to 13 across the board. This is irrelevant for
Nixpkgs as we pin a specific LLVM version anyway, and versions below
LLVM 18 will be removed imminently. Secondly, support for the hidden
`-fno-llvm-tbaa` flag is dropped. This can be replaced with custom
`-optlo` flags to control the passes more directly, but the main
use of this undocumented flag appears to have been to [work around]
the lack of support for newer LLVM versions, anyway.

[work around]: <https://gitlab.haskell.org/ghc/ghc/-/issues/22220>

I successfully built the following on `aarch64-linux`:

* `pkgsCross.armv7l-hf-multiplatform.buildPackages.haskell.compiler.ghc948`
* `pkgsCross.armv7l-hf-multiplatform.buildPackages.haskell.compiler.ghc967`
* `pkgsCross.armv7l-hf-multiplatform.buildPackages.haskell.compiler.ghc984`
* `pkgsCross.armv7l-hf-multiplatform.buildPackages.haskell.compiler.ghc9102`
* `pkgsCross.armv7l-hf-multiplatform.buildPackages.haskell.compiler.ghc9121`
* `pkgsCross.armv7l-hf-multiplatform.buildPackages.haskell.compiler.ghc9122`
* `pkgsCross.riscv64.haskell.compiler.ghc948`

The GHC 9.4.8 with an ARMv7 host platform segfaults when I try to run
GHC, though e.g. `ghc-pkg --help` runs successfully. The GHC 9.10.3
build targeting ARMv7 crashed inside `llc(1)`, so I tried RISC‐V,
which has some platform mismatch issue relating to `libffi`, so I
tried z/Architecture, which failed with an invalid floating point
constant in the LLVM IR, so I tried 64‐bit MIPS, which failed with
a different `libffi` issue, so I tried 32‐bit MIPS, which failed
to compile `compiler-rt`, so I gave up. I confirmed that both of the
ARMv7 issues reproduce with 944e8fd4f45f214f99e908eac4a69162cb9a4196,
the revision before they were bumped from their old versions of LLVM,
so these are not regressions.

I built a test program with the ARMv7 cross‐compilers and
confirmed that they run on the AArch64 builder. I also confirmed
that the cross‐compiled RISC‐V GHC successfully runs under
`qemu-riscv64(1)`. It will only try to build programs via the C
backend, though, as that is the only option for unregisterised™
targets, so it’s not clear to me how useful LLVM support in 9.4.8
really is for bootstrapping new platforms; I guess even RISC‐V
would require more backporting work to produce a cross‐compiled
GHC that will use LLVM to compile its own input. I didn’t bother
setting up all the binfmt machinery to get it through compiling and
running a test program, but it at least makes the attempt.

(cherry picked from commit b6be8a03a76d7b58d613389b6d99c92fc7c1488d)

wolfgangwalther committed:
haskell.compiler.ghc963Binary: drop

This bootstraps mostly from ghc984Binary, at least for the major
platforms. The bootstrap paths for darwin become much shorter.
Unfortunately, the bootstrap paths for i686-linux become a bit longer,
so it's a trade-off.

(cherry picked from commit 5df6c9a28f1ddcb54ca5f558e0764ab92e782904)

emilazy committed:
haskell.packages.ghc902Binary: strip down for bootstrap

This should only be used to bootstrap source‐built GHCs.

emilazy committed:
haskell.compiler.ghc924Binary: drop

(cherry picked from commit c9c734465795ce42e8c9eb39aa8f815340d3ca5e)

emilazy committed:
haskell.compiler.ghc967: bootstrap with GHC 9.4.8

We’d want every version to be able to trace back to 9.4.8 for the
“cross‐compiled GHC bootstrap tarballs” plan, anyway. I’m not
sure whether we’d want a `ghc948Binary` package that also includes
the official tarballs, or just only use our cross‐compiled binary
distributions, but it will make sense to align the version used for
bootstrap regardless.

(cherry picked from commit da915f81efab5163f7a3a3fd304970f07909b336)

ghpzin committed:
haskell.compiler.ghc{948,967,984,9101,9121,9122}: fix build with gcc15

- add patch from fedora (rebase of latest fix from upstream):
https://gitlab.haskell.org/ghc/ghc/-/issues/25662
https://gitlab.haskell.org/ghc/ghc/-/merge_requests/13863

ghc versions excluded from patch:
9.10.2 - has previous fix with `extern void* malloc(size_t);`
9.10.3 - has latest fix with `#include <stdlib.h>`
9.12.3 - has latest fix with `#include <stdlib.h>`

Fixes build failure with gcc15, building utils/hp2ps/Utilities.c:
```
utils/hp2ps/Utilities.c:6:14: error:
warning: conflicting types for built-in function ‘malloc’;
expected ‘void *(long unsigned int)’ [-Wbuiltin-declaration-mismatch]
        6 | extern void* malloc();
          |              ^~~~~~
...
utils/hp2ps/Utilities.c:92:18: error:
warning: conflicting types for built-in function ‘realloc’;
expected ‘void *(void *, long unsigned int)’ [-Wbuiltin-declaration-mismatch]
       92 |     extern void *realloc();
          |                  ^~~~~~~
```

(cherry picked from commit 02746ead92b6d56d02fb22efacd59402d1a95878)

wolfgangwalther committed:
compilers/ghc/common-hadrian: remove unused condition

Hadrian is only used for GHC 9.6+ anyway.

(cherry picked from commit 14488d4051243696786742844af1ebe908c9f36c)

alexfmpe committed:
haskellPackages: fix executable builds on pkgsCross.ucrt64

alexfmpe committed:
haskellPackages.network: fix build on windows

alexfmpe committed:
haskellPackages.streaming-commons: fix build on windows

sternenseemann committed:
haskell.compiler.ghc984Binary: drop unnecessary libnuma dep

ghc902Binary's rts.conf indeed prescribes linking against numa for
aarch64-linux, but this is not the case with the later bindists we have
been providing numa for.

(cherry picked from commit 347062bef5fb63ffe30742a52ca89540d925e8d5)

emilazy committed:
haskell.compiler.ghc948: use `autoreconfHook`

This matches the Hadrian build, and will be required for the LLVM
support backports.

(cherry picked from commit 6664f3137c2ee106bb6ea12eb7b88e0382cad50c)

emilazy committed:
haskell.compiler.ghc948: drop obsolete `configure` patch

This string is not present in GHC 9.4.8’s `configure` script.

(cherry picked from commit 7d28e619b43cff32a2f1bcab8875c4b3a3819ac8)

emilazy committed:
haskell.compiler.ghc948: don’t patch both `aclocal.m4` and `configure`

`autoreconfHook` will handle the latter now.

(cherry picked from commit 5923f1f1e2d266d6d59954afa87012a1746d36f3)

wolfgangwalther committed:
haskellPackages: stackage LTS 24.12 -> LTS 24.15

all-cabal-hashes: 2025-09-25T14:09:07Z -> 2025-10-15T11:31:05Z

(generated by maintainers/scripts/haskell/update-package-set.sh)

emilazy committed:
haskellPackages: remove dead code for removed GHC versions

emilazy committed:
haskellPackages: drop packages only used for removed GHC versions

emilazy committed:
haskellPackages: drop unused patch file

nixpkgs-ci[bot] committed:
Merge 7d4c6daffb7270bc95399325b1c1ca67356deaf9 into haskell-updates

wolfgangwalther committed:
haskellPackages: clean up after removals (#452393)

sternenseemann committed:
haskellPackages.pandoc: update test fixture for djot >= 0.1.2.3

sternenseemann committed:
haskellPackages.hs-opentelemetry-api: drop released patches

wolfgangwalther committed:
haskellPackages.pandoc: update test fixture for djot >= 0.1.2.3 (#452699)

nixpkgs-ci[bot] committed:
Merge 5d1f6df5e189570d6a24bbb80cb5731a63fa41d2 into haskell-updates

wolfgangwalther committed:
haskellPackages.digit: remove outdated override

Setup.lhs has now been deleted upstream.

nixpkgs-ci[bot] committed:
Merge a1e822ab3bdcabe44059f627b820c4d1cec5827b into haskell-updates

sternenseemann committed:
git-annex: update sha256 for 10.20250929

dotlambda committed:
hledger: 1.43.2 -> 1.50.2

sternenseemann committed:
Merge commit 257aa2b0468e689341d21a0f36be855d08 into haskell-updates

sternenseemann committed:
haskellPackages.cabal2nix-unstable: 2025-09-17 -> 2025-10-19

haskellPackages.aes-gcm: unbreak

https://github.com/NixOS/cabal2nix/pull/673

nixpkgs-ci[bot] committed:
Merge 275782673cefcdad61a446c3ab94312213b52562 into haskell-updates

wolfgangwalther committed:
hledger: 1.43.2 -> 1.50.2 (#453350)

nixpkgs-ci[bot] committed:
Merge d08251c5835c82e0e1c4c49f1692293fe8ab13bc into haskell-updates

wolfgangwalther committed:
haskellPackages: stackage LTS 24.15 -> LTS 24.16

all-cabal-hashes: 2025-10-15T11:31:05Z -> 2025-10-22T21:20:29Z

(generated by maintainers/scripts/haskell/update-package-set.sh)

nixpkgs-ci[bot] committed:
Merge f1d065ea2710d39a4b99ee739357ef7fcbf6bf13 into haskell-updates

nixpkgs-ci[bot] committed:
Merge 22bb9b307966c009c1bf1df746190833e8989e98 into haskell-updates

wolfgangwalther committed:
haskellPackages: stackage LTS 24.15 -> LTS 24.16 (#453822)

wolfgangwalther committed:
haskellPackages.proto3-wire: remove merged patches

alexfmpe committed:
haskellPackages.Win32: mark as boot package

alexfmpe commented:
> We have a lot of the following eval errors on the branch right now:

`git bisect` landed on the single commit from https://github.com/NixOS/nixpkgs/pull/366593.
Reverting it indeed fixes eval again, and builds/executables seem to work as well.

A quick glance at `--show-trace` suggests we have this chain
  ghcjs -> emscripten -> openjdk -> python -> openssl

culminating at

https://github.com/NixOS/nixpkgs/blob/be6d38e0740004012d82791b907112f2564bd926/pkgs/development/libraries/openssl/default.nix#L118

which then throws.

I guess it makes sense we don't have a C compiler on ghcjs 'platform', but why is this chain getting evaluated as ghcjs anyway?

wolfgangwalther committed:
haskellPackages: more pkgsCross.ucrt64 fixes (#449004)

nixpkgs-ci[bot] committed:
Merge c9f97ce3914379dedf762d10e08d665d1f550d10 into haskell-updates

nixpkgs-ci[bot] committed:
Merge e32661f795c4a156189fe911082cc1892ea04091 into haskell-updates

sternenseemann committed:
haskellPackages.mkDerivation: make license optional

`stdenv.mkDerivation` does not require meta.license to be passed, so there
is no reason `haskellPackages.mkDerivation` needs to enforce this. This
would free up cabal2nix to not report a license if it is not sure.

As I have argued in https://github.com/NixOS/cabal2nix/pull/677#issuecomment-3444635379,
it is better not to report a license than reporting an inaccurate one.

This would also allow to stop generating arbitrary strings as licenses
in cabal2nix to remove string values to facilitate more cleanliness in
the spirit of https://github.com/NixOS/nixpkgs/pull/445672, though the
question is whether it is wise to remove the meta data altogether.

sternenseemann committed:
haskellPackages.cabal2nix-unstable: 2025-10-19 -> 2025-10-25

ners committed:
haskell: fix hls-plugin-api, ghcide, lsp-types ambiguity

jerith666 committed:
haskellPackages.ghci-dap: unbreak

this was unbroken by the update from 0.0.25.0 to 0.0.26.0 in
743815bde61e, which added support for ghc >= 9.10.2.

sternenseemann committed:
haskellPackages.mkDerivation: make license optional (#455796)

wolfgangwalther committed:
haskellPackages.cabal2nix-unstable: 2025-10-19 -> 2025-10-25 (#455808)

nixpkgs-ci[bot] committed:
Merge 2e8ab34df47b2543166294457a4288b8d89b4e17 into haskell-updates

wolfgangwalther committed:
haskellPackages.ghci-dap: unbreak (#455973)

maralorn committed:
haskell-language-server: Fix build

maralorn committed:
haskell: fix hls-plugin-api, ghcide, lsp-types ambiguity (#455884)

wolfgangwalther committed:
haskellPackages: unbreak various packages

Created by maintainers/scripts/haskell/unbreak.nu.

wolfgangwalther committed:
Merge commit 'ba840a89e061c8b91d00daa7d5c3da5f4a96936d' into haskell-updates

maralorn committed:
haskell-language-server: Fix build (#456151)

wolfgangwalther committed:
postgrest: 13.0.7 -> 14.0

Release Notes:
https://github.com/PostgREST/postgrest/releases/tag/v14.0

wolfgangwalther committed:
postgrest: 13.0.7 -> 14.0 (#456429)

nixpkgs-ci[bot] committed:
Merge 64f174a74b6591145d772d941170acb1893c1829 into haskell-updates

nixpkgs-ci[bot] committed:
Merge d229de74ad65c3e25c09dde2bde3e3af250994ea into haskell-updates

sternenseemann committed:
haskellPackages.cabal2nix-unstable: 2025-10-25 -> 2025-10-31

sternenseemann committed:
haskellPackages.nix-paths: provide pkgs.nix for libraryToolDepends

sternenseemann committed:
haskellPackages.nix-paths:don't use build tools when cross compiling

Using build-tools is somewhat flawed since it relies on the host tools
being executable at build time which is not always the case, e.g. when
cross-compiling. The simplest solution (patching being the alternative)
is to simply use the flag that makes nix-paths use PATH at runtime.

sternenseemann committed:
haskellPackages.nix-paths: fix after upload of `nix` hackage package (#457300)

nixpkgs-ci[bot] committed:
Merge 92661251e05809a81dbcc341cba3c9dd4317ec44 into haskell-updates

nixpkgs-ci[bot] committed:
Merge 2a8c7895537748fb6d8f0a14c08a482ffe549441 into haskell-updates

sternenseemann committed:
release-haskell.nix: reflect reorganized agda tests

maralorn committed:
haskell.packages.ghc94.haskell-language-server: Add deprecation warning

maralorn commented:
Re: markdown-unlit in hls (9.6 + 9.8¥. This is because of a new tutorial executable that we we have no business building. I just need to figure out how to disable building it.

sternenseemann committed:
haskellPackages.dom-parser: fix build with text >= 2.1.2

sternenseemann committed:
haskellPackages: mark builds failing on hydra as broken

This commit has been generated by maintainers/scripts/haskell/mark-broken.sh based on
*evaluation [1819781](https://hydra.nixos.org/eval/1819781) of nixpkgs commit [52d502c](https://github.com/NixOS/nixpkgs/commits/52d502c55197c8adce03a73bc01a617f7543c286) as of 2025-11-02 21:08 UTC*
from the haskell-updates jobset on hydra under https://hydra.nixos.org/jobset/nixpkgs/haskell-updates

maralorn committed:
haskell.packages.ghc94.haskell-language-server: Add deprecation warning (#457892)

sternenseemann committed:
haskellPackages.pandoc-crossref: pin to 0.3.21

wolfgangwalther commented:
> (Please don’t bypass the merge queue in the future, it’s there for a reason!)

Unfortuantely, because of how things are set up, it's essentially impossible to merge haskell-updates with the merge queue, because GitHub always shows conflicts.

(that being said, it is possible to run eval locally after creating the merge conflict)

I will need to wake up first and then look into the Eval failure soon-ish.
